src/methods/properties.js

```diff
@@ -1100,15 +1116,23 @@ export class PropertiesImplementation {
           (current: any)[field] = (Desc: any)[field];
         }
       }
-      InternalUpdatedProperty(realm, O, P, oldDesc);
+      InternalEmitPropertyMutationEffect(realm, Target, P, current, oldDesc);
     }
 
     // 11. Return true.
     return true;
   }
 
   // ECMA262 9.1.6.1
-  OrdinaryDefineOwnProperty(realm: Realm, O: ObjectValue, P: PropertyKeyValue, Desc: Descriptor): boolean {
+  // Note: The extra Target argument represents the Value that should receive any emitted effects,
+  // or null if none should be emitted.
```

Comment:
We still have to update the internal property binding so that we know what the new value is when we try to read out of it - which is a complicated set of rules.

https://github.com/facebook/prepack/blob/fbc32cc76deeb0735c9a58cb0ad1e55269bea47d/src/values/ArrayValue.js#L207

I could maybe add an alternative method name but I didn't want people to try to use it without a target and forget to add the target when there should be one. Maybe if it had a really discouraged name. In the end, I'd probably still want to pass null for this instead of forking the internals of this method.

It's subtle since the standard method doesn't have this.

I considered adding a method just for emitting the effects, but it turned out to be very complicated because we have to read the previous state of the binding to determine if we can use assignment and emit the effect after figuring out the final descriptor etc. It also made it easy to forget to do this in new places, the Target argument forces us to think about it.

In the end I couldn't figure out a better structure.